File: core/src/main/java/com/views/ldebugview/LDebugViewService.java
package com.views.ldebugview;

import android.annotation.SuppressLint;
import android.app.Service;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.os.Build;
import android.os.IBinder;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.R;
import com.core.utilities.LAppResource;
import com.core.utilities.LDateUtil;
import com.core.utilities.LScreenUtil;
import com.core.utilities.LUIUtil;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

public class LDebugViewService extends Service implements View.OnTouchListener {
    private final String logTag = getClass().getSimpleName();
    private WindowManager mWindowManager;
    private View mFloatingView;
    private WindowManager.LayoutParams params;
    private View collapsedView;
    private View expandedView;
    private LinearLayout llRootTv;
    private ScrollView scrollView;

    public LDebugViewService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        EventBus.getDefault().register(this);

        //Inflate the floating view layout we created
        mFloatingView = LayoutInflater.from(this).inflate(R.layout.layout_debug_view_service, null);

        //Add the view to the window.
        int LAYOUT_FLAG;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            LAYOUT_FLAG = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY;
        } else {
            LAYOUT_FLAG = WindowManager.LayoutParams.TYPE_PHONE;
        }
        params = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                LAYOUT_FLAG,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT);

        //Specify the view position
        params.gravity = Gravity.TOP | Gravity.LEFT;        //Initially view will be added to top-left corner
        params.x = 0;
        params.y = 100;

        //Add the view to the window
        mWindowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        mWindowManager.addView(mFloatingView, params);

        collapsedView = mFloatingView.findViewById(R.id.rlCollapse);
        expandedView = mFloatingView.findViewById(R.id.llExpanded);

        expandedView.getLayoutParams().width = LScreenUtil.Companion.getScreenWidth() / 2;
        expandedView.getLayoutParams().height = LScreenUtil.Companion.getScreenHeight() * 2 / 3;
        expandedView.requestLayout();

        llRootTv = mFloatingView.findViewById(R.id.ll_root_tv);
        scrollView = mFloatingView.findViewById(R.id.scrollView);

        //Set the close button
        ImageView closeButtonCollapsed = mFloatingView.findViewById(R.id.ivClose);
        closeButtonCollapsed.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //close the service and remove the from from the window
                stopSelf();
            }
        });

        //Set the close button
        ImageView closeButton = mFloatingView.findViewById(R.id.ivCloseButton);
        closeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                collapsedView.setVisibility(View.VISIBLE);
                expandedView.setVisibility(View.GONE);
            }
        });

        //Drag and move floating view using user's touch action.
        mFloatingView.findViewById(R.id.rlRootContainer).setOnTouchListener(this);
    }

    /**
     * Detect if the floating view is collapsed or expanded.
     *
     * @return true if the floating view is collapsed.
     */
    private boolean isViewCollapsed() {
        return mFloatingView == null || mFloatingView.findViewById(R.id.rlCollapse).getVisibility() == View.VISIBLE;
    }

    private int initialX;
    private int initialY;
    private float initialTouchX;
    private float initialTouchY;

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        if (v.getId() == R.id.rlRootContainer) {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:

                    //remember the initial position.
                    initialX = params.x;
                    initialY = params.y;

                    //get the touch location
                    initialTouchX = event.getRawX();
                    initialTouchY = event.getRawY();
                    return true;
                case MotionEvent.ACTION_UP:
                    int Xdiff = (int) (event.getRawX() - initialTouchX);
                    int Ydiff = (int) (event.getRawY() - initialTouchY);

                    //The check for Xdiff <10 && YDiff< 10 because sometime elements moves a little while clicking.
                    //So that is click event.
                    if (Xdiff < 10 && Ydiff < 10) {
                        if (isViewCollapsed()) {
                            //When user clicks on the image view of the collapsed layout,
                            //visibility of the collapsed layout will be changed to "View.GONE"
                            //and expanded view will become visible.
                            collapsedView.setVisibility(View.GONE);
                            expandedView.setVisibility(View.VISIBLE);
                        }
                    }
                    return true;
                case MotionEvent.ACTION_MOVE:
                    //Calculate the X and Y coordinates of the view.
                    params.x = initialX + (int) (event.getRawX() - initialTouchX);
                    params.y = initialY + (int) (event.getRawY() - initialTouchY);

                    //Update the layout with new X & Y coordinate
                    mWindowManager.updateViewLayout(mFloatingView, params);
                    return true;
            }
            return false;
        }
        return false;
    }

    @SuppressLint("SetTextI18n")
    private void print(LComunicateDebug.MsgFromActivity msgFromActivity) {
        if (msgFromActivity == null) {
            return;
        }
        String currentTime = LDateUtil.Companion.getDateCurrentTimeZoneMls(System.currentTimeMillis(), "HH:mm:ss");
        TextView textView = new TextView(this);
        LUIUtil.Companion.setTextSize(textView, LAppResource.INSTANCE.getDimenValue(R.dimen.txt_medium));

        if (msgFromActivity.getObject() == null) {
//            Log.d(logTag, "msgFromActivity.getObject() == null");
            textView.setText(currentTime + " : " + msgFromActivity.getMsg());
        } else {
//            Log.d(logTag, "msgFromActivity.getObject() != null");
            LUIUtil.Companion.printBeautyJson(msgFromActivity.getObject(), textView, currentTime);
        }

        LUIUtil.Companion.setTextSize(textView, getBaseContext().getResources().getDimension(R.dimen.text_small));
        if (msgFromActivity.getType() == LComunicateDebug.MsgFromActivity.TYPE_D) {
            textView.setTextColor(Color.WHITE);
        } else if (msgFromActivity.getType() == LComunicateDebug.MsgFromActivity.TYPE_E) {
            textView.setTextColor(Color.RED);
        } else if (msgFromActivity.getType() == LComunicateDebug.MsgFromActivity.TYPE_I) {
            textView.setTextColor(Color.GREEN);
        }
        llRootTv.addView(textView);
        scrollView.post(new Runnable() {
            public void run() {
                scrollView.fullScroll(View.FOCUS_DOWN);
            }
        });
    }

    @Override
    public void onDestroy() {
        EventBus.getDefault().unregister(this);
        if (mFloatingView != null) {
            mWindowManager.removeView(mFloatingView);
        }
        super.onDestroy();
    }

    //listen msg from activity
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(LComunicateDebug.MsgFromActivity msg) {
        print(msg);
    }
}

File: core/src/main/java/com/loitpcore/views/lDebugView/LDebugViewService.kt
package com.loitpcore.views.lDebugView

import android.annotation.SuppressLint
import android.app.Service
import android.content.Intent
import android.graphics.Color
import android.graphics.PixelFormat
import android.os.Build
import android.os.IBinder
import android.view.* // ktlint-disable no-wildcard-imports
import android.view.View.OnTouchListener
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import com.loitpcore.R
import com.loitp.core.utilities.LAppResource
import com.loitp.core.utilities.LDateUtil
import com.loitp.core.utilities.LScreenUtil
import com.loitp.core.utilities.LUIUtil
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class LDebugViewService : Service(), OnTouchListener {

    private var mWindowManager: WindowManager? = null
    private var mFloatingView: View? = null
    private var params: WindowManager.LayoutParams? = null
    private var collapsedView: View? = null
    private var expandedView: View? = null
    private var llRootTv: LinearLayout? = null
    private var scrollView: ScrollView? = null

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    @SuppressLint("InflateParams")
    override fun onCreate() {
        super.onCreate()
        EventBus.getDefault().register(this)

        mFloatingView = LayoutInflater.from(this).inflate(R.layout.layout_debug_view_service, null)

        val layoutFlag: Int = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
        } else {
            @Suppress("DEPRECATION")
            WindowManager.LayoutParams.TYPE_PHONE
        }
        params = WindowManager.LayoutParams(
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            layoutFlag,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        )

        // Specify the view position
        params?.let {
            it.gravity =
                Gravity.TOP or Gravity.START // Initially view will be added to top-left corner
            it.x = 0
            it.y = 100
        }

        // Add the view to the window
        mWindowManager = getSystemService(WINDOW_SERVICE) as WindowManager
        mWindowManager?.addView(mFloatingView, params)

        collapsedView = mFloatingView?.findViewById(R.id.rlCollapse)
        expandedView = mFloatingView?.findViewById(R.id.llExpanded)

        expandedView?.apply {
            layoutParams.width = LScreenUtil.screenWidth / 2
            layoutParams.height = LScreenUtil.screenHeight * 2 / 3
            requestLayout()
        }

        llRootTv = mFloatingView?.findViewById(R.id.ll_root_tv)
        scrollView = mFloatingView?.findViewById(R.id.scrollView)

        // Set the close button
        val closeButtonCollapsed = mFloatingView?.findViewById<ImageView>(R.id.ivClose)
        closeButtonCollapsed?.setOnClickListener {
            // close the service and remove the from from the window
            stopSelf()
        }

        // Set the close button
        val closeButton = mFloatingView?.findViewById<ImageView>(R.id.ivCloseButton)
        closeButton?.setOnClickListener {
            collapsedView?.visibility = View.VISIBLE
            expandedView?.visibility = View.GONE
        }

        // Drag and move floating view using user's touch action.
        mFloatingView?.findViewById<View>(R.id.rlRootContainer)?.setOnTouchListener(this)
    }

    /**
     * Detect if the floating view is collapsed or expanded.
     *
     * @return true if the floating view is collapsed.
     */
    private val isViewCollapsed: Boolean
        get() = mFloatingView == null || mFloatingView?.findViewById<View>(R.id.rlCollapse)?.visibility == View.VISIBLE
    private var initialX = 0
    private var initialY = 0
    private var initialTouchX = 0f
    private var initialTouchY = 0f

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(v: View, event: MotionEvent): Boolean {
        if (v.id == R.id.rlRootContainer) {
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {

                    // remember the initial position.

                    params?.let {
                        initialX = it.x
                        initialY = it.y

                        // get the touch location
                        initialTouchX = event.rawX
                        initialTouchY = event.rawY
                    }

                    return true
                }
                MotionEvent.ACTION_UP -> {
                    val xDiff = (event.rawX - initialTouchX).toInt()
                    val yDiff = (event.rawY - initialTouchY).toInt()

                    // The check for Xdiff <10 && YDiff< 10 because sometime elements moves a little while clicking.
                    // So that is click event.
                    if (xDiff < 10 && yDiff < 10) {
                        if (isViewCollapsed) {
                            // When user clicks on the image view of the collapsed layout,
                            // visibility of the collapsed layout will be changed to "View.GONE"
                            // and expanded view will become visible.
                            collapsedView?.visibility = View.GONE
                            expandedView?.visibility = View.VISIBLE
                        }
                    }
                    return true
                }
                MotionEvent.ACTION_MOVE -> {
                    // Calculate the X and Y coordinates of the view.
                    params?.x = initialX + (event.rawX - initialTouchX).toInt()
                    params?.y = initialY + (event.rawY - initialTouchY).toInt()

                    // Update the layout with new X & Y coordinate
                    mWindowManager?.updateViewLayout(mFloatingView, params)
                    return true
                }
            }
            return false
        }
        return false
    }

    @SuppressLint("SetTextI18n")
    private fun print(msgFromActivity: LComunicateDebug.MsgFromActivity?) {
        if (msgFromActivity == null) {
            return
        }
        val currentTime =
            LDateUtil.getDateCurrentTimeZoneMls(System.currentTimeMillis(), "HH:mm:ss")
        val textView = TextView(this)
        LUIUtil.setTextSize(textView, LAppResource.getDimenValue(R.dimen.txt_medium).toFloat())
        if (msgFromActivity.any == null) {
            textView.text = currentTime + " : " + msgFromActivity.msg
        } else {
            LUIUtil.printBeautyJson(o = msgFromActivity.any, textView = textView, tag = currentTime)
        }
        LUIUtil.setTextSize(
            textView = textView,
            size = baseContext.resources.getDimension(R.dimen.txt_tiny)
        )
        when (msgFromActivity.type) {
            LComunicateDebug.MsgFromActivity.TYPE_D -> {
                textView.setTextColor(Color.WHITE)
            }
            LComunicateDebug.MsgFromActivity.TYPE_E -> {
                textView.setTextColor(Color.RED)
            }
            LComunicateDebug.MsgFromActivity.TYPE_I -> {
                textView.setTextColor(Color.GREEN)
            }
        }
        llRootTv?.addView(textView)
        scrollView?.let { sv ->
            sv.post {
                sv.fullScroll(View.FOCUS_DOWN)
            }
        }
    }

    override fun onDestroy() {
        EventBus.getDefault().unregister(this)
        mWindowManager?.removeView(mFloatingView)
        super.onDestroy()
    }

    // listen msg from activity
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(msg: LComunicateDebug.MsgFromActivity?) {
        print(msgFromActivity = msg)
    }
}
